[#23] 판매자 로그인 스케일아웃

2eve2 opened the pull request.

2eve2 committed:
feat: seller 회원가입, 로그인 등

- user api refactoring
- loginProcess 중복 -> service 통합 고려

2eve2 committed:
fix: user - seller login conflict 문제 해결

2eve2 committed:
refactor: login interceptor

- 명구님 redisSessionInterceptor에서 @LoginCheck 확인 로직 추가 후 기존 loginCheckInterceptor 제거
- interceptor 설정에서 redisInterceptor 체크 후 logincheckInterceptor로 넘어가는데, redisInterceptor에서 이미 토큰 - 세션 정보 확인 후 세션을 저장하는데 logincheckInterceptor에서 한번 더 세션 유무 여부를 확인하지 않아도 된다고 생각해서 지움.
- redisInterceptor가 @LoginCheck가 없어도 작동되어 미인증 오류를 전달함해서 관련 로직 추가함

2eve2 committed:
feat: session tracking mode 추가
- URL 전달 방식을 끄고 항상 쿠키를 통해서만 세션을 유지

2eve2 committed:
refactor: redis 분리

- +) LoginCheckInterceptor 다시 추가함.

2eve2 committed:
refactor: redis 분리 yml

2eve2 committed:
refactor: interceptor 로직

2eve2 committed:
refactor: login interceptor

- role 구분
- RedisSessionLoginInterceptor 삭제
- ErrorCode 수정 등

2eve2 committed:
fix: checkstyle

2eve2 committed:
refactor: httpSession 생성시점 변경, optional 추가

2eve2 committed:
refactor: 로그아웃 프로세서 분리

2eve2 committed:
feat: 세션 redis

- 하나의 redis session storage에 user, seller session 모두 저장되는 문제 해결

2eve2 committed:
chore: checkstyle

2eve2 committed:
refactor: redis 통합

- user, seller의 redis session storage를 분리하지 않고 통합함. user 대비 seller의 수가 많지 않을 것이며 서비스 초기에는 사용자 수가 많지 않은데 분리하여 관리할 필요성이 적다고 생각함. 사용자 수가 많아지면 분리 고려.

2eve2 committed:
feat: session properties 추가

2eve2 committed:
feat: session logout

2eve2 committed:
refactor: user, seller 권한 인증

2eve2 committed:
chore: 개행 수정

dding94 commented on common/src/main/java/com/flab/common/auth/AuthenticatedSeller.java:
```diff
@@ -0,0 +1,62 @@
+    private Role role;
+    private LocalDateTime expireAt;
+
+    public AuthenticatedSeller() {
+    }
```
해당 부분 public 보다는 protected 로 선언하는게 실수도 방지하고, 안전하게 사용할 수 있다는 생각이 듭니다..!

2eve2 replied:
앗..! 네 이 부분 빠뜨렸네요. 수정하겠습니다

dding94 commented on seller-api/seller-application/src/main/java/com/flab/seller/application/CreateSellerProcessor.java:
```diff
@@ -0,0 +1,33 @@
+
+        String encryptedPassword = passwordEncryptor.encrypt(command.getPassword());
+        return this.sellerRepository.save(command.toEntity(encryptedPassword));
+    }
```
해당부분에서 this는 빼는게 어떨까요? 현재 저도 구현할때 프로세서에서도 this 사용하지 않아서 맞추면 괜찮지 않을까 합니다.

dding94 commented on seller-api/seller-application/src/main/java/com/flab/seller/application/SearchSellerProcessor.java:
```diff
@@ -11,7 +12,7 @@ public SearchSellerProcessor(SellerRepository sellerRepository) {
     }
 
-    public Seller execute(Long id) {
+    public Optional<Seller> execute(Long id) {
```
repository 에서 옵셔널 반환안되게 처리하는 방법은 어떻게 생각하시나요? 뭐가 더 좋은 방법인지는 모르겠지만 저는 repository 에서 그렇게 처리했습니다!

2eve2 replied:
옵셔널 반환이 다른 레이어까지 이어지지 않아서 좀 더 깔끔해보이는 장점이 있을까요? 우선 옵셔널 처리를 미리 해서 넘기는 방향으로 바꾸겠습니다.

dding94 commented on seller-api/seller-infrastructure/src/main/java/com/flab/seller/infrastructure/config/SellerConfig.java:
```diff
@@ -0,0 +1,71 @@
+    public CreateSellerProcessor createSellerProcessor(
+            SellerRepository sellerRepository
+    ) {
+        return new CreateSellerProcessor(
+                sellerRepository,
+                new SellerSecurityPasswordEncoder(sellerEncodingAlgorithm()));
+    }
```
해당 부분 체크스타일이 좀 이상하네요 인코딩 알고리즘 옆에 괄호 하나 내리면 될 것 같습니다.

dding94 commented on seller-api/seller-infrastructure/src/main/java/com/flab/seller/infrastructure/config/SellerConfig.java:
```diff
@@ -0,0 +1,71 @@
+    public SearchSellerProcessor searchSellerProcessor(
+            SellerRepository sellerRepository
+    ) {
+        return new SearchSellerProcessor(
+                sellerRepository
+        );
```
이부분은 파라미터가 짧아서 엔터처리 안해도 될 것 같다는 생각이 듭니다

dding94 commented on seller-api/seller-infrastructure/src/main/java/com/flab/seller/infrastructure/persistence/redis/RedisSessionRepository.java:
```diff
@@ -0,0 +1,71 @@
+
+    private long calculateExpirationTime(AuthenticatedSeller authenticatedSeller) {
+        long sessionExpirationSec = sessionProperties.getSessionExpirationSec();
+        long sellerExpirationSec = redisTemplate.getExpire(
+                authenticatedSeller.getSessionId(),
+                TimeUnit.SECONDS
+        );
```
해당부분 주의창 나오지 않으신가요?? 제가 만든 부분도 그렇네요. 아직 해결방법이 보이진 않아서 저도 찾아보겠습니다.

2eve2 replied:
확인해보니 wrapper 클래스를 primitive 로 그냥 변환하는 과정에서 null check를 하지 않아 발생하는 경고창이었네요. null check를 하는 조건 추가하면 해결될 것 같습니다. 명구님 파트도 수정해놓겠습니다!

2eve2 replied:
우선 null 체크 조건 추가했는데 DataNotFoundException 클래스를 따로 만들어서 redis errorcode를 따로 만들까하다가 필요성을 크게 못느껴서 IllegalArgumentException()으로 처리해두었습니다. 다른 의견 있으면 남겨주세요 :)

2eve2 committed:
refactor: 리뷰 반영

- checkstyle
- Unboxing null 체크 등

f-lab-bright commented on common/build.gradle:
```diff
@@ -1,3 +1,12 @@
+plugins {
+    id 'io.spring.dependency-management' version '1.0.12.RELEASE'
+    id 'org.springframework.boot' version '2.7.2'
```
common에는 요 플러그인 설정도 안해도 되지 않나요?

2eve2 replied:
앗 네 common에서 플러그인 설정 하지 않아도 되는걸 몰랐었네요. 수정하겠습니다.

f-lab-bright commented on seller-api/seller-application/src/main/java/com/flab/seller/application/LoginSellerProcessor.java:
```diff
@@ -1,15 +1,66 @@
-    public LoginSellerProcessor(SellerRepository sellerRepository) {
+
+    public LoginSellerProcessor(
+            SellerRepository sellerRepository,
+            PasswordEncryptor passwordEncryptor,
+            SessionRepository sessionRepository,
+            Long sessionExpirationSec, HttpSession session
+    ) {
         this.sellerRepository = sellerRepository;
+        this.passwordEncryptor = passwordEncryptor;
+        this.sessionRepository = sessionRepository;
+        this.sessionExpirationSec = sessionExpirationSec;
+        this.session = session;
     }
 
-    //todo login 구현
+    @Transactional
+    public String execute(LoginSellerCommand command) {
+        var seller = sellerRepository.findByEmail(command.getEmail());
+
+        if (!passwordCheck(command, seller)) {
+            throw new SellerPasswordNotMatchedException();
+        }
+
+        var loginSellerInfo = seller.toLoginInfo();
+
+        // 세션 정보 생성
+        session.setAttribute(AUTH_SESSION_MEMBER, loginSellerInfo);
+        session.setAttribute(AUTH_STATUS, Role.SELLER);
+
+        // redis session 정보 저장
+        loginSellerInfo.addSessionInfo(session.getId(), sessionExpirationSec);
+        sessionRepository.save(session.getId(), loginSellerInfo);
+
+        return session.getId();
+    }
```
session이라는 개념을 application layer가 알고 있어야 할까요?
jwt 같은 토큰 형태로 바뀌게 되면 해당 코드가 통채로 바뀌어야 하겠네요.
session / token 등의 개념은 presentation layer에서 인증/인가을 어떻게 처리할지에 대한 책임이지 않을까요?

2eve2 replied:
나중에 로그인 인증 방식이 변경될 가능성을 생각지 못하고 구현했던 것 같습니다.
application layer에서는 session, token 등의 개념을 알 필요 없이 아이디, 패스워드 체크에 통과된 멤버 정보를 넘겨주고, presentation layer에서 따로 인증처리를 할 수 있도록 변경하겠습니다.

다만 세션 생성 및 세션 정보를 세션 스토리지(레디스)에 저장하는 것을 직접 presentation layer에서 코드로 작성하는 것보다는 나중에 jwt 등 방식으로 바뀔 경우를 생각하여 인증 서비스 클래스를 따로 만드는 것이 좋을 것 같다고 생각했는데요. 구현 후 코드리뷰 요청 다시 드리겠습니다.

f-lab-bright commented on server/src/main/java/com/flab/livecommerce/auth/LoginCheckInterceptor.java:
```diff
@@ -0,0 +1,89 @@
+
+        Role authority = loginCheck.authority();
+
+        switch (authority) {
+            case USER:
+                checkUserAuthority(request);
+                break;
+
+            case SELLER:
+                checkSellerAuthority(request);
+                break;
+
+            default:
+                throw new AuthenticationException();
+        }
```
타입이 생길 때 마다 switch 문을 추가하고, 인증처리 하는 코드를 매번 다시 작성하는건 그다지 좋아보이는 구조는 아닌 것 같긴 하네요

2eve2 replied:
이 부분에 대해서 고민을 해보았는데 Role 도메인에서 권한 valid 체크를 하고 interceptor 내에서는 if문으로 AuthenticatedUser, AuthenticatedSeller를 체크하도록 변경해보았습니다. switch문은 없어졌지만 인증 처리해주는 부분은 User와 Seller의 Authenticated 객체가 달라서 if문으로 나눠놨는데 좋은 구조인지는 잘 모르겠네요..

dding94 replied:
이 구조도 switch 와 마찬가지로 타입이 생길 때 마다 동일하게 If 문이 추가가 될 것 같습니다. 해당 방법 보다는 타입을 순회해서 찾는 경우로 만들면 타입이 추가가 되어도 나머지는 변경이 없도록 코드를 만들 수 있을 것 같습니다.

2eve2 replied:
> LGTM with Comments 롤이 또 추가되었을 때도 충분히 적은 코드로 추가될 수 있을지 고민해보시면 좋을 것 같네요. 병합하고 진행해보시죠 :)

현재 stream으로 Role 안에서 타입을 순회해서 Role 요소가 변경되더라도 문제 없게끔 리팩토링하였습니다 :) 병합하겠습니다!

2eve2 committed:
refactor: 리뷰 반영

- common build.gradle
- LoginCheckInterceptor switch문 -> 도메인 책임 전가

2eve2 committed:
refactor: 인증 서비스 분리 (틀 완성, 진행중)

dding94 commented on common/src/main/java/com/flab/common/auth/Role.java:
```diff
@@ -1,5 +1,18 @@
-    USER, SELLER;
+    UNAUTH, USER, SELLER;
+
+    public void valid(Role memberRole) {
+        if (this == Role.USER && this != memberRole) {
+            throw new AuthorizationException(ErrorCode.USER_AUTHORIZATION);
+        }
+        if (this == Role.SELLER && this != memberRole) {
+            throw new AuthorizationException(ErrorCode.SELLER_AUTHORIZATION);
+        }
+    }
+
```
해당부분은 Role 이 추가될 때 마다 if 문이 계속 늘어나는 구조가 될 것 같습니다. 다른 방법도 있지 않을까요?

2eve2 committed:
refactor: AuthenticatedUser, Seller 객체 통합

- 같은 정보를 가지고 있는데 분리하여 인터셉터 로직이 불필요하게 복잡해짐.
- 추후 인증/인가 방식이 변경되는 것을 고려했을 때 token, sessionId로 변수 명칭을 지정해두고 사용하는 것은 확장을 고려하지 않은 구조라고 판단.
- WARNING: 추후 병합 시 충돌 우려.

2eve2 committed:
Merge branch 'main' into feature/23-seller-login-scaleout

2eve2 committed:
refactor: Builder 패턴 적용

- 생성자로서 필요한 파라미터가 통일되지 않아 여러 메서드가 사용되었고, 빌더 패턴으로 리팩토링

2eve2 committed:
Merge remote-tracking branch 'origin/feature/23-seller-login-scaleout' into feature/23-seller-login-scaleout

2eve2 committed:
fix: 병합 문제 해결

f-lab-bright reviewed (approved):
LGTM with Comments
롤이 또 추가되었을 때도 충분히 적은 코드로 추가될 수 있을지 고민해보시면 좋을 것 같네요.
병합하고 진행해보시죠 :)

2eve2 committed:
refactor: Role 변경에 유연한 구조로 변경

- 기존) if 문으로 타입 지정 -> Role 추가 시 코드 추가해야 함.
- 변경) stream으로 타입 순회 후 검증 로직 -> Role 추가가 되더라도 코드 변경이 필요 없게 리팩토링

2eve2 committed:
refactor: Seller Login Response 추가

- client측의 편의를 위해 jsessionid 리턴하도록 변경

2eve2 commented:
인증/인가 처리 방식이 User, Seller가 다른 상황에서 추후 인증 방식의 변경될 여지를 고려해 구조를 전반적으로 리팩토링하였습니다. 병합 시 충돌을 최소화하기 위해 UserApi 부분은 로직이 묶이는 부분만 리팩토링하였는데 추후 다른 pr에서 User Api 나머지 파트도 반영하겠습니다.
